config_converter/config_mapper/config_mapper.py

```diff
@@ -0,0 +1,198 @@
+        'rotate_wait': 'rotate_wait'
+    }
+}
+# each plugin (key) maps to a list of all possible special fields (value)
+# these fields may not belong to the same level, not have a 1:1 mapping, etc
+# https://docs.fluentd.org/parser/multiline - shows formatN works for
+# 1 <= N <= 20
+_SPECIAL_MAP = {
+    'in_tail': [
+        'format', 'format_firstline', '@type', 'multiline_flush_interval',
+        'expression'
+    ] + [f'format{i}' for i in range(1, 21)]
+}
+# each main plugin type (key) maps to a dict of required fields (value)
+# these sub-dicts are common for all plugins of a type, hence different keys
+# each sub-dict maps fluentd field names (key) to master agent fields (value)
+# each sub-dict only contains fields that go in the upper level after mapping
+_OUTER_LEVEL_MAP = {'source': {'tag': 'name', '@type': 'type'}}
+# each plugin (key) maps to a list of all possible directives (value)
+# each list contains known and allowed sub directives
+_SUPPORTED_DIRS = {'in_tail': ['parse']}
+# list of fields we do not know how to convert
+_UNSUPPORTED_FIELDS = [
+    'emit_unmatched_lines', 'enable_stat_watcher', 'enable_watch_timer',
+    'encoding', 'from_encoding', 'ignore_repeated_permission_error',
+    'limit_recently_modified', 'open_on_every_update', 'path_timezone',
+    'pos_file_compaction_interval', 'read_from_head', 'read_lines_limit',
+    'skip_refresh_on_startup'
+]
+# plugins we know how to convert
+_SUPPORTED_PLUGINS = ['in_tail']
+# plugins we do not know how to convert
+_UNSUPPORTED_PLUGINS = ['in_forward', 'in_syslog']
+
+
+def extract_root_dirs(config_obj: config_pb2.Directive) -> dict:
+    """Checks root dirs, maps with corresponding params if supported."""
+    result = {'logs_module': dict()}
+    plugin_prefix_map = {'source': 'in_', 'match': 'out_'}
+    dir_name_map = {'source': 'sources', 'match': 'output'}
+    # these dicts can be updated when more plugins are supported
+    for d in config_obj.directives:
+        if d.name not in plugin_prefix_map:
+            print("Currently we do not support plugins of " + d.name)
+            continue
+        for p in d.params:
+            if p.name == '@type':
+                # identify the plugin
+                plugin = plugin_prefix_map[d.name] + p.value
+                if plugin in _SUPPORTED_PLUGINS:
+                    result['logs_module'][dir_name_map[d.name]] = \
+                            result['logs_module'].get(
+                                dir_name_map[d.name], [])
+                    converted_dir: dict = _plugin_convert(d, plugin)
+                    result['logs_module'][dir_name_map[d.name]]\
+                                .append(converted_dir)
+                elif plugin in _UNSUPPORTED_PLUGINS:
+                    print("we dont support plugin " + plugin)
+                else:
+                    print("we dont know plugin " + plugin)
+    return result
+
+
+def _plugin_convert(d: config_pb2.Directive, plugin: str) -> dict:
```

Comment:
I think it would be clearer to start with a less-general solution here.

I would suggest starting with a separate function for each plugin we support (so far just in_tail).  Something like:

    def convert_in_tail(d: config_pb2.Directive) -> dict:
      fields = {}
      if has_param(d, 'pos_file'):
        fields['checkpoint_file'] = get_param(d, 'pos_file')
      if has_param(d, 'path'):
        fields['path'] = get_param(d, 'path')
      ...
      return {'file_source_config': fields}

What do you think?

Reply:
Discussed over call - will follow this template.